### PR TITLE
[WIP] Develop fixing promise to Develop

### PR DESCRIPTION
**Pull request by Radiateurs**
Still need to modify DataController, SelectorUtils and availableField.
Need to add descriptions headers

### CODE REVIEW

File: db/migrations/20180621153556_users-table.js [hunk 1]
```diff
@@ -2,7 +2,7 @@ exports.up = function (knex) {
     return knex.schema.createTable('USERS', function (table) {
         table.increments('id').primary();
         table.text('username').notNullable();
-        table.text('realname');
+        table.text('realname').nullable();
```
Comment (fguitton) [inline db/migrations/20180621153556_users-table.js]: We can open a ticket about this eventually, but I think we should make sure prople have a real name all the time. There is no regulatory reason we'd have to not do it and it is much nicer to provide reports

File: src/controllers/ceController.js [hunk 2]
```diff
@@ -1,37 +1,54 @@
...
-            res.status(400).send('Missing visit id');
-        }
+function CeController() {
+    this.clinicalEvent = new clinicalEventCore();
+
+    this.createCe = CeController.prototype.createCe.bind(this);
+    this.deleteCe = CeController.prototype.deleteCe.bind(this);
+}
+
+CeController.prototype.createCe = function(req, res) {    //need to change
```
Comment (fguitton) [inline src/controllers/ceController.js]: Can we understand better what "need[s] to change" ?

File: src/controllers/patientController.js [hunk 3]
```diff
@@ -1,183 +1,217 @@
...
+    this.patient = new PatientCore();
+
+    this.searchPatients = PatientController.prototype.searchPatients.bind(this);
+    this.createPatient = PatientController.prototype.createPatient.bind(this);
+    this.setPatientAsDeleted = PatientController.prototype.setPatientAsDeleted.bind(this);
+    this.getPatientProfileById = PatientController.prototype.getPatientProfileById.bind(this);
+    this.erasePatientInfo = PatientController.prototype.erasePatientInfo.bind(this);
+}
+
+PatientController.prototype.searchPatients = function (req, res) {  //get all list of patient if no query string; get similar if querystring is provided
```
Comment (fguitton) [inline src/controllers/patientController.js]: Lets make sure we bring all that in a proper JSDoc format at some point

File: src/controllers/patientController.js [hunk 4]
```diff
@@ -1,183 +1,217 @@
...
-                                    eraseEntry(res, 'VISITS', { 'id': visitId[i] }, 'Row for visits', null, false);
-                                }
+/**
+ * TODO : Patch flow
+ */
+PatientController.prototype.erasePatientInfo = function (req, res) {
+    let patientId = undefined;
+    let visitId = [];
+    if (req.requester.priv !== 1) {
+        res.status(403).send('Sorry! Only admins are able to edit / delete data');
```
Comment (fguitton) [inline src/controllers/patientController.js]: There are still multiple message not in a locale file. We need to keep track of that.

File: src/controllers/testController.js [hunk 5]
```diff
@@ -1,57 +1,90 @@
...
+TestController.prototype.addActualOccurredDate = function (req, res) {
+    if (req.body.hasOwnProperty('testId') && req.body.hasOwnProperty('actualOccurredDate')) {
+        this.test.addActualOccurDateTest(req.requester, { id: req.body.testId, actualOccurDate: req.bosy.actualOccurDate }).then(function (result) {
+            res.status(200).json(result);
+            return;
+        }, function (error) {
+            res.status(400).json(ErrorHelper(message.errorMessages.UPDATEFAIL, error));
+            return;
+        });
+        // knex('ORDERED_TESTS')
```
Comment (fguitton) [inline src/controllers/testController.js]: I'd like that we go over all the commented code at some point and understand what needs to be kept or not. Is there a reason for this commented code to still be around ?

File: src/utils/db-handler.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+const knex = require('./db-connection');
+const fs = require('fs');
+
+function migrate(type){
+    switch (type) {
+        case 'testing':
+            knex.migrate.latest({ directory: './db/migrations' })
+                .then(() => knex.seed.run({ directory: './db/seed' }))
+                .then(() => knex.seed.run({ directory: './db/exampleDataForTesting/seed' }))
+                .then(() => { if (process.env.NODE_ENV !== 'PROD') console.log('CREATED DATABASE AND POPULATED WITH MS MODULES AND TESTING DATA'); })
```
Comment (fguitton) [inline src/utils/db-handler.js]: `PROD` is not a vaild/conventional verb for `NODE_ENV`. Use `production` instead. Also, in this file there are multiple duplications which would benefit from an extracted message handler.

File: src/utils/db-handler.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+                .then(() => knex.destroy())
+                .catch(err => { console.log(err); knex.destroy(); });
+            break;
+        case  'MS_fields':
+            knex.migrate.latest({ directory: './db/migrations' })
+                .then(() => knex.seed.run({ directory: './db/seed' }))
+                .then(() => { if (process.env.NODE_ENV !== 'PROD') console.log('CREATED DATABASE AND POPULATED WITH MS MODULES'); })
```
Comment (fguitton) [inline src/utils/db-handler.js]: `PROD` is not a vaild/conventional verb for `NODE_ENV`. Use `production` instead. Also, in this file there are multiple duplications which would benefit from an extracted message handler.

File: src/utils/db-handler.js [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+                .then(() => knex.destroy())
+                .catch(err => { console.log(err); knex.destroy(); });
+            break;
+        case 'bare':
+            knex.migrate.latest({ directory: './db/migrations' })
+                .then(() => { if (process.env.NODE_ENV !== 'PROD') console.log('CREATED DATABASE WITH BARE SCHEMA'); })
```
Comment (fguitton) [inline src/utils/db-handler.js]: `PROD` is not a vaild/conventional verb for `NODE_ENV`. Use `production` instead. Also, in this file there are multiple duplications which would benefit from an extracted message handler.

File: src/controllers/testController.js (hunk 5 above)
Comment (Radiateurs) [reply src/controllers/testController.js]: Fixed in f3de6b7495e274da4044214c8d161249418fd5a0

File: src/controllers/patientController.js [hunk 9]
```diff
@@ -1,183 +1,217 @@
...
-                    });
-            });
-        res.status(200).send('Erasure completed. Check for any data retreivable if needed.');
-    }
-}
-
+                    }
+                });
+        }).catch(err => {
+            if (process.env.NODE_ENV !== 'prod')
```
[the remark this answers is not in the document]
Comment (Radiateurs) [reply src/controllers/patientController.js]: Fixed in 300e0f72acd3f119ee6a78ecb5c14416203053ca

File: src/server.js [hunk 10]
```diff
@@ -1,3 +1,11 @@
 const app = require('./app');
 
-app.listen(3000, () => { console.log('listening on port 3000!'); });
\ No newline at end of file
+let server = app.listen(3000, () => { console.log('listening on port 3000!'); });
+
+process.on('exit', function(){
+    server.close(function (){
+        if (process.env.NODE_ENV !== 'PROD')
```
[the remark this answers is not in the document]
Comment (Radiateurs) [reply src/server.js]: Fixed in 300e0f72acd3f119ee6a78ecb5c14416203053ca

File: src/utils/db-handler.js (hunk 6 above)
Comment (Radiateurs) [reply src/utils/db-handler.js]: Fixed in 300e0f72acd3f119ee6a78ecb5c14416203053ca

File: src/utils/db-handler.js (hunk 7 above)
Comment (Radiateurs) [reply src/utils/db-handler.js]: Fixed in 300e0f72acd3f119ee6a78ecb5c14416203053ca

File: src/utils/db-handler.js (hunk 8 above)
Comment (Radiateurs) [reply src/utils/db-handler.js]: Fixed in 300e0f72acd3f119ee6a78ecb5c14416203053ca

**Pull request merged**